crates/runix/src/flake_ref2/mod.rs

```diff
@@ -0,0 +1,267 @@
+use std::borrow::Cow;
+use std::fmt::Display;
+use std::str::FromStr;
+
+use chrono::{NaiveDateTime, TimeZone, Utc};
+use derive_more::{Display, From};
+use serde::{Deserialize, Deserializer, Serialize};
+use thiserror::Error;
+
+use self::file::FileRef;
+use self::git::GitRef;
+use self::git_service::{service, GitServiceSource};
+use self::indirect::IndirectRef;
+use self::path::PathRef;
+
+pub mod file;
+pub mod git;
+pub mod git_service;
+pub mod indirect;
+pub mod lock;
+pub mod path;
+pub mod protocol;
+
+pub trait FlakeRefSource: FromStr + Display {
+    fn scheme() -> Cow<'static, str>;
+    fn parses(maybe_ref: &str) -> bool {
+        maybe_ref.starts_with(&format!("{}:", Self::scheme()))
+    }
+}
+
+#[derive(Serialize, Deserialize, Display, From, Debug)]
+#[serde(untagged)]
+pub enum FlakeRef {
+    FileFile(FileRef<protocol::File, file::File>),
+    FileHTTP(FileRef<protocol::HTTP, file::File>),
+    FileHTTPS(FileRef<protocol::HTTPS, file::File>),
+    TarballFile(FileRef<protocol::File, file::Tarball>),
+    TarballHTTP(FileRef<protocol::HTTP, file::Tarball>),
+    TarballHTTPS(FileRef<protocol::HTTPS, file::Tarball>),
```

Comment:
What's the advantage of making each of these a different type? I haven't gone through in enough detail yet to see if we're handling them all very differently, but I'm wondering why not `Tarball { protocol: enum Protocol { File, HTTP, HTTPS } }`

Reply:
@ysndr thoughs on this?

Reply:
Discussed and may be worth reworking at some point but not now